lib/ddtrace/writer.rb

```diff
@@ -95,10 +94,12 @@ def write(trace, services)
       # This check ensures that if a process doesn't own the current +Writer+, async workers
       # will be initialized again (but only once for each process).
       pid = Process.pid
-      @mutex_after_fork.synchronize do
-        if pid != @pid
-          # we should start threads because the worker doesn't own this
-          start()
+      if pid != @pid
+        @mutex_after_fork.synchronize do
+          if pid != @pid
```

Comment:
Yes - the first comparison is done outside of the mutex to avoid locking as this check is only True once per the lifetime of the worker.

The second comparison is to make sure that other thread didn't already start the worker in between first check and locking of the mutex.

Reply:
If this is the case, I think it might be a good idea to add comments explaining this, at each line which it appears. As it's written, it's a bit unintuitive, so such an explanation would be really helpful.

Reply:
Yeah comment with explanation sounds like a good idea.

BTW @delner  this could be done a bit cleaner with `AtomicFixnum` from `concurrent-ruby` but we would have to add it as a dependency.

Reply:
My feelings are that it isn't worth adding a dependency for making things a bit cleaner, and that the tracing library should be as thin as reasonably possible.